File: Python Datastructures/Singly Linked List/SinglyLinkedList.py
from typing import Generic
from typing import TypeVar

#None is null in python

T = TypeVar("T")

#Define the node class 
class Node(Generic[T]):
  def  __init__ (self, data: T):
    self.data = data
    self.nxt  = None

  
 

class SinglyLinkedList:
  # A singly linked list uses a single pointer to reference the
  # Next node
  def __init__(self) -> None:
    self.size: int = 0
    self.head: Node = None
    self.tail: Node = None

  #Indexer for the singly linked list
  def __getitem__ (self, index: int) -> Node:
    #Base case to check if the index is larger then the size of the singly
    #linked list or if the list is empty
    if(index > self.size - 1 or self.isEmpty()):
        raise Exception("Out of range")
      
    #Set the current item to the head   
    currentItem = self.head
    #Loop over the singly linked list using a range from the index
    for n in range(index):
      #Set the current item
      currentItem = currentItem.nxt
    
    #Set the currentItem
    return currentItem
  
  #ToString handling
  def __str__(self):
    #Set the output we will be adding to
    output = "["

    # Get the current item
    currentItem = self.head

    #Loop through the list using a range from the current size of the singly
    #linked list
    for n in range(self.size):
      #Add to the value to the output
      output += f"{currentItem.data}"
      #If the currentItems nxt node is not None then we will also add a comma 
      if(currentItem.nxt != None):
        output += ","
      #Set the current item
      currentItem = currentItem.nxt
    #Add the closing bracket to the string.
    output += "]"
    #Return the string
    return output

  # Add Element
  def add(self, item: Node) -> None:
    # Call the add last element method
    self.addLast(item)

  # Add Element to the end of the list
  def addLast(self, item: Node) -> None:
    #If the list is empty then we will set the head and tail values
    if(self.isEmpty()):
      self.head = item
      self.tail = item
    #Else set the tail's next value to the item passed in, then set the tail to
    #the item passed in 
    else:
      self.tail.nxt = item
      self.tail = item
    #Add to the size of the list
    self.size += 1

  # Add Element to the begin of the list
  def addFirst(self, item: Node) -> None:
    #If the list is empty then we will set the head and tail values
    if(self.isEmpty()):
      self.head = item
      self.tail = item
    #Else set the items, next value to the head then set the header to the item
    #passed in
    else:
      item.nxt = self.head
      self.head = item
    #Add to the size of the list
    self.size += 1

  # Remove the first value from the SinglyLinked List
  def removeFirst(self) -> None:
    #Set a header ref value for garbage collection if needed. This will be
    #handled behind the scene in python
    headRef = None # for garbage collection if needed.
    #If the singly linked list is empty then throw an exception 
    if(self.isEmpty()):
        raise Exception("This linked list is empty")
    #If we only have a single value then set the head and tail to null
    elif(self.size == 1):
      self.head = None
      self.tail = None
    #Else set the header to the headers next node
    else:
      self.head = self.head.nxt
    # subtract from the size
    self.size -= 1

  # Remove the last value from the Singlylinked List
  def removeLast(self) -> None:
    #If the signly linked list is empty then throw and exception
    if(self.isEmpty()):
        raise Exception("This linked list is empty")
    #If the singly linked list only has on value then set the head and tail to None
    elif(self.size == 1):
      self.head = None
      self.tail = None
    #Else iterate throught the singly linked list and set the nodes to remove
    #the last node
    else:
      #Set a tail ref incase we need to garbage collect.
      tailNodeRef = self.tail # For garbage collection if needed
      #Set a placeholder for the previous Node
      previousNode = None
      #Set the current value we are looping over
      currentNode = self.head

      #Loop though the until we find a currentItem where the next value is None
      while currentNode.nxt is not None:
        #Set the previousItem to the current item
        previousNode = currentNode
        #Set the currentItem to the next node
        currentNode = currentNode.nxt

    #Set the previousItems next node to None
    previousNode.nxt = None
    #Set the tail
    self.tail = previousNode
    #Subtract from the size
    self.size -= 1
      
  #Remove an item by the data of a node    
  def removeItem(self, data: T) -> None:
    #If the singly linked list is empty throw an exception
    if(self.isEmpty()):
      raise Exception("This linked list is empty")
    #If the head.data equals the data passed in then set the head to the current
    #heads next value
    elif(self.head.data == data):
      self.head = self.head.nxt
      self.size -= 1
      return
    
    #If the size is only a length of 1 then clear the head and tail
    elif(self.size == 1):
      self.head = None
      self.tail = None
    else:
      
      #Set the NodRef Placeholder for garbage collection
      nodeRef = None #for garbage collection
      #Set the previous for the previous node
      previousNode = None
      #Set the current node
      currentNode = self.head
      
      #Loop though the nodes until we get to a node that is none
      while currentNode is not None:
        #If the currentNode's data matches the data passed in then break the loop
        if(currentNode.data == data):
            break
        #Set the previous node to the current node
        previousNode = currentNode
        #Set the current node to the next node of the current node
        currentNode = currentNode.nxt
    #If the current node is none then we will return
    if(currentNode == None):
      return

    #Set the previous node's next node to the current node's next value    
    previousNode.nxt = currentNode.nxt
    #Subtract from the size of the singly linked list
    self.size -= 1
    

  #Check if the linked list is empty.
  def isEmpty(self) -> bool:
    #Return the bool value from the check
    return self.size == 0
  
  #Method to get the index
  def indexOf(self, data:T) -> int:
    # Set the index
    index = 0
    
    #Set the current item to the head value
    currentItem = self.head
    #Loop throught the items while the current items is not None
    while currentItem is not None:
      # If the current items data equals the data passed in then break from the loop
      if(currentItem.data == data):
        break
      # Set the current item to the next item the current item has a point to
      currentItem = currentItem.nxt
      # Increase the size of the index
      index += 1
    # If the index is larger then the size of the linked list then we will
    # return -1
    if(index > self.size):
      return -1
    # Else we will return the index
    else:
      return index
    
  #Method to check if the singly linked list contains a node with certain data
  def contains(self, data: T) -> bool:
    # Get the index of the data passed in. If the index returns -1 then we will
    # return false. Else we will return true 
    return self.indexOf(data) != -1

 
#Create list
list = SinglyLinkedList()
#Add values to the list
list.add(Node("Mon"))
list.add(Node("Tues"))
list.add(Node("Wed"))
list.add(Node("Thurs"))
list.add(Node("Fri"))
list.add(Node("Sat"))
list.add(Node("Sun"))


#Method Testing

File: C# DataStructures/Datastructures/SinglyLinkedList/SinglyLinkedList.cs
using C__DataStructures.Datastructures.Shared.Base;
using C__DataStructures.Datastructures.Shared.Nodes;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;

namespace C__DataStructures.Datastructures.SinglyLinkedList
{
   
    public class SinglyLinkedList<T> : BaseLinkedList<OneWayNode<T>,T>
    {
        public SinglyLinkedList()
        {
            Length = 0;
            Head = null;
        }

        public SinglyLinkedList(OneWayNode<T> item)
        {
            Length = 1;
            Head = item;
        }
        public override OneWayNode<T> this[int index] { 
            get {

                if (index > Length - 1 || index < 0|| IsEmpty()) { 
                    throw new IndexOutOfRangeException("Out of range. Index is two large or two small.");
                }

                var toReturn = Head;

                for (int i = 0; i <= index; i++) {
                    toReturn = toReturn.Next;
                }

                if (toReturn == null) {
                    throw new NullReferenceException("Incountered a null reference issue while tring to retrive the item by the index");
                }

                return toReturn;
            }
        }

        public override void Add(OneWayNode<T> item)
        {
            AddLast(item);
        }

        protected internal override void AddLast(OneWayNode<T> item)
        {

            //If the list if empty and the head is null. Then we will set the head of the linked list
            //Else we will loop through the linked list and add to the end of it
            //Time Complexity: O(N)
            if (IsEmpty() || Head == null)
            {
                Head = item;
                Length++;
            }
            else
            {
                //set the place holder for the previous node
                OneWayNode<T>? prevNode = null;
                //set the place holder for the current node
                OneWayNode<T>? currentNode = Head;
                //Loop throught the nodes and get to the end of the list
                while (currentNode != null)
                {
                    prevNode = currentNode;
                    currentNode = currentNode.Next;
                }

                //set the previous nodes next pointer value
                prevNode.Next = item;
                //add to the length
                Length++;
            }
        }

        public override void AddFirst(OneWayNode<T> item)
        {
            //if the list is empty or if the header is null. Then we will set the head item
            //Else we will create a reference to the head value, then we will set the new head as the item passed in an we will set its next pointer
            if (IsEmpty() || Head == null)
            {
                Head = item;
            }
            else { 
                OneWayNode<T> currentHeadNode = Head;

                Head = item;

                item.Next = currentHeadNode;
            }

        }

        public override void Clear()
        {
            //set the length to zero
            Length = 0;
            //clear the header which will drop all of the values on the linked list.
            Head = null;
            
        }

        public override OneWayNode<T>? Remove(OneWayNode<T> item)
        {
            //Call the protected method to remove the items and return the removed item
           return RemoveItem(item);
        }

        public override OneWayNode<T> RemoveFirst()
        {
            //Check if the list is empty. If it is throw an error
            if (IsEmpty() || Head == null)
            {
                throw new InvalidOperationException("The Linked list is currently empty. Please add a value to it before continuing.");
            }
            //else clear out the head node
            else {
                //Set a placeholder var for the head to return
                var headPlaceHolder = Head;
                //Clear the head node.
                Head = Head.Next;
                //subtract from the length.
                Length--;
                //Return the placeholder
                return headPlaceHolder;
            }
        }

        public override OneWayNode<T> RemoveLast()
        {
            //If the list is empty and the head is null then we will throw an error
            if (IsEmpty() || Head == null)
            {
                //Throw an error
                throw new InvalidOperationException("The Linked list is currently empty. Please add a value to it before continuing.");
            }
            //If the length is 1 and the head is not null then we will set the head to null and subtract from the length
            else if (Length == 1 && Head != null)
            {
                //Create a place holder for the current head value.
                var headPlaceHolder = Head;

                //Set the head to null
                Head = Head.Next;

                //Subtract from the length
                Length--;

                //Return the head placeholder.
                return headPlaceHolder;
                //Subtract from the length of the list
                Length--;
            }
            //Else traverse through the list, until we get to the end, keeping track of the value before the tail node.
            //Then set the second to last node next pointer to null
            else
            {
                //Set the placeholder for the prev value
                OneWayNode<T>? prevNode = null;
                //Set the place holder for the current value 
                OneWayNode<T>? currentNode = Head;

                //This may be incorrect for getting the last node. Need to figure out a proper way of getting the last node without a counter
                while (currentNode.Next != null)
                {
                    prevNode = currentNode;
                    currentNode = currentNode.Next;
                }

                //Set the next node to null.
                prevNode.Next = null;
     
                return prevNode;
            }
        }
        protected internal override OneWayNode<T>? RemoveItem(OneWayNode<T> item)
        {
            //Check if the list is empty or null. If it is then throw an error
            if (IsEmpty() || Head == null)
            {
                throw new InvalidOperationException("The Linked list is currently empty. Please add a value to it before continuing.");
            }
            //Check if the lenght of the list is one, the head is not null and the head data matches the 
            else if (Head != null && Comparer<T>.Default.Compare(item.Data, Head.Data) >= 0)
            {
                //Set a placeholder for the head value
                var headPlaceHolder = Head;
                //Change the head to is next value. This will either be the next node or null.
                Head = Head.Next;
                //Subtract from the length 
                Length--;
                //return the palceholder
                return headPlaceHolder;
            }
            else {

                //Set the placeholder for the prev value
                OneWayNode<T>? prevNode = null;
                //Set the place holder for the current value 
                OneWayNode<T>? currentNode = Head.Next;

                //Loop though until we find the node we are looking for or until the currentnode is null
                while (currentNode != null) {
                    //Break from the while loop early if we find the item we are looking for
                    if (Comparer<T>.Default.Compare(item.Data, currentNode.Data) >= 0) {
                        break;
                    }

                    //Set the prev node to the current node
                    prevNode = currentNode;
                    //Set the current node to the current nodes next node
                    currentNode = currentNode.Next;
                }

                //If the the current node is null then throw an error
                if (currentNode == null) {
                    throw new ArgumentException("Invalid item pass to this method to remove. No item was found to remove");
                }

                //Set the prev nodes next node to the node to removes next node
                prevNode.Next = currentNode.Next;

                //Subtract from the length
                Length--;

                //return the current node.
                return currentNode;
            }
        }

        public override bool Contains(OneWayNode<T> item)
        {
            if (IsEmpty() || Head == null)
            {
                throw new InvalidOperationException("The Linked list is currently empty. Please add a value to it before continuing.");
            }
            else {
                for (int index = 0; index <= Length - 1; index++) {
                    if (Comparer<T>.Default.Compare(item.Data, this[index].Data) >= 0) {
                        return true;
                    }
                }    
                return false;
            }
            
        }

        public override int IndexOf(OneWayNode<T> item)
        {

            if (IsEmpty() || Head == null)
            {
                throw new InvalidOperationException("The Linked list is currently empty. Please add a value to it before continuing.");
            }
            else
            {
                for (int index = 0; index <= Length - 1; index++)
                {
                    if (Comparer<T>.Default.Compare(item.Data, this[index].Data) >= 0)
                    {
                        return index;
                    }
                }
               return -1;
            }
        }

        public override bool IsEmpty()
        {
            //if the length is less than or equal to zero then we will return true, else it will return false.
            return Length <= 0;
        }

        public override string ToString()
        {
            throw new NotImplementedException();
        }

    }
}
